pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,455 @@
+		Hostname:     machine.Spec.Name,
+		UserData:     userdata,
+		ProjectID:    c.ProjectID,
+		Facility:     c.Facilities,
+		BillingCycle: c.BillingCycle,
+		Plan:         c.InstanceType,
+		OS:           imageName,
+		Tags: []string{
+			generateTag(string(machine.UID)),
+		},
+	}
+
+	device, res, err := client.Devices.Create(serverCreateOpts)
+	if err != nil {
+		return nil, packetErrorToTerminalError(err, res, "failed to create server")
+	}
+
+	return &packetDevice{device: device}, nil
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData) (bool, error) {
+	instance, err := p.Get(machine)
+	if err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	res, err := client.Devices.Delete(instance.(*packetDevice).device.ID)
+	if err != nil {
+		return false, packetErrorToTerminalError(err, res, "failed to delete the server")
+	}
+
+	return false, nil
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine) (instance.Instance, error) {
+	device, _, err := p.getPacketDevice(machine)
+	if err != nil {
+		return nil, err
+	}
+	if device != nil {
+		return &packetDevice{device: device}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, newID types.UID) error {
+	device, client, err := p.getPacketDevice(machine)
+	if err != nil {
+		return err
+	}
+
+	// go through existing labels, make sure that no other UID label exists
+	tags := make([]string, 0)
+	for _, t := range device.Tags {
+		// filter out old UID tag(s)
+		if _, err := getTagUID(t); err != nil {
+			tags = append(tags, t)
+		}
+	}
+
+	// create a new UID label
+	tags = append(tags, generateTag(string(newID)))
+
+	glog.Infof("Setting UID label for machine %s", machine.Name)
+	dur := &packngo.DeviceUpdateRequest{
+		Tags: &tags,
+	}
+	_, response, err := client.Devices.Update(device.ID, dur)
+	if err != nil {
+		return packetErrorToTerminalError(err, response, "failed to update UID label")
+	}
+	glog.Infof("Successfully set UID label for machine %s", machine.Name)
+
+	return nil
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) MachineMetricsLabels(machine *v1alpha1.Machine) (map[string]string, error) {
+	labels := make(map[string]string)
+
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err == nil {
+		labels["size"] = c.InstanceType
+		labels["facilities"] = strings.Join(c.Facilities, ",")
```

Comment:
Not completely happy about this, as it doesn't actually tell the user in which facility the instance ended up being if there was more than one submitted. But I am not sure what a better solution is, as doing an API call every time `MachineMetricsLabels` gets called is definitely too expensive

Reply:
Right. It is an interesting feature of Packet's API that you can specify one, multiple or "any" facility. See [here](https://www.packet.com/developers/api/#create-a-device):

```
The facilities attribute specifies in what datacenter you wish to create the device.

You can either specify a single facility { "facility": "f1" } , or you can instruct to create the device in the best available datacenter { "facility": "any" }. Additionally it is possible to set a prioritized location selection.

For example { "facility": ["f3", "f2", "any"] } will try to assign to the facility f3, if there are no available f2, and so on. If "any" is not specified for "facility", the request will fail unless it can assign in the selected locations.
```

When you actually do `Create()` the device, it returns a `device` as well. I don't know if that returns the _requested_ facilities or _actual_ facility. I have some go code with a quick experiment to find out. 

If it does, could we not just retrieve it, since we save the device at the end of `Create()`?

Reply:
Confirmed. The return from the call _does_ tell you what facility it is in, and we save that:

```go
func (p *provider) Create(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData, userdata string) (instance.Instance, error) {
        // ...
        device, res, err := client.Devices.Create(serverCreateOpts)  
        // This "device" contains it in `device.Facility.Name`
        // ...
        // and we return it here
        return &packetDevice{device: device}, nil
}
```

Reply:
Its not that easy, because we would need to store this state somewhere, probably in `.Status.ProviderStatus`. We can leave that as a follow-up thought.

Reply:
Makes sense. We do _have_ it, though, so that is a good first step. It is easier to find a place to store something we have, than to find a way to get that which we do not. :-)